### DISCUSSION TITLE
Performance benchmarks

### DISCUSSION THREAD

**Discussion by Almouro** [reactwg/react-native-new-architecture#85] [Q&A] [answered]
Hi, thanks for creating this space! ❤️

I'm currently running some performance benchmarks comparing old arch and new arch and the results I'm getting are surprising to me.

I'm using 
- a low end Samsung J3 2017 Android 9 (on `armeabi-v7a`) for the videos 
- a low end Samsung A10s Android 10 (on `armeabi-v7a`) on AWS device farm for the measures

## Scenario 1: comparing rendering 10k views

**WINNER** New arch ✅

The winner is clearly the new arch here and by far. Kudos on that! 🤩

https://user-images.githubusercontent.com/4534323/196978522-56d40532-54ff-4430-b93e-490afa704a64.mp4

I'm also running some perf metrics comparison using our own tool (https://github.com/bamlab/android-performance-profiler) that's basically a glorified `adb shell top`

The [full report](https://rn-new-arch-perf.netlify.app/a10s/manyviews/report) here makes it clear that this is because the new renderer is much more efficient than the old bridge

![image](https://user-images.githubusercontent.com/4534323/196980050-65b22956-1a8a-44e0-9a97-87a1bad07729.png)

## Scenario 2: comparing rendering 2k texts

**WINNER** Old arch ❌

https://user-images.githubusercontent.com/4534323/196980411-6c5d8889-161f-4735-ab2c-9e321ac6f622.mp4

This time the old arch is faster, mainly because of heavier UI thread consumption 

![image](https://user-images.githubusercontent.com/4534323/196980783-ef21ecba-9c36-44df-8285-6cff565bed91.png)

All the code [is here](https://github.com/Almouro/fabric-test).

This might have a big impact on performance of apps migration:
- I can reproduce similar results in a [3rd scenario](https://twitter.com/almouro/status/1583070070275985409?s=20&t=2YnDRboLcOn_5BlqsRFrKA), displaying some Tweets components
- This might be related to performance issues mentioned on the Flashlist migration to Fabric https://github.com/Shopify/flash-list/pull/550

I haven't tried to go deeper with a system trace to check what was going on yet, and my understanding of how the new arch works and implements layouting is limited still. 

But I'm hoping to get a discussion started towards improving performance (and/or the benchmarks) and what could be done, if someone already has some ideas ☺️

**Comment by psionic12** [comment]
Same here, I tested the `rn-tester` and the new arch is about 50ms slower than the old arch, after testing piece by piece, it seems that the layout procedure in new arch takes more time. I'm not sure if it is because of the performance of the code in UIManger, or the new arch somehow make the view hierarchy more complicated than before.

**Comment by sammy-SC** [comment, marked as answer]
Hello Almouro,

First of all, thank you very much for sharing your benchmarks, as well as a playground where we could look into and understand what is happening.

I’ve personally tried to reproduce your results on both a Pixel 5 and a Pixel 3 physical device. I ran many-texts and 100-tweets scenarios. On both devices the CPU performances were about neutral. We’re aware of the higher memory consumption. The new architecture has complete in memory representation of UI. This adds the extra memory cost but it unlocks new features like concurrent rendering. We’re looking into ways to optimise memory usage.

![311424609_1514233375682787_7775965648106729546_n](https://user-images.githubusercontent.com/1733610/198001400-73cbc5ca-184c-42ea-9e5a-5baa6c6968b3.png)


To better understand what would be the culprit here, I’d like to ask follow up questions:

1. Have you tried running the benchmarks on more devices?
2. Is there a specific reason why you used a device farm (AWS) for measurements?
3. Have you tried comparing performance on iOS? 


I’d like to take the opportunity to say that the React Native team is fully committed to rollout the New Architecture of React Native, without introducing regressions for our users. I’m personally interested in taking a closer look at any well-reported regression.

To give you a bit of context, performance was a big concern during the rollout of the New Architecture in our internal products. Our internal performance evaluation was neutral across all React Native surfaces in the Facebook app on both Android and iOS. In the last years, we conducted dozens of tests in production on millions of devices to assure performance was neutral. Still, we’re open to looking into benchmarks like the one you reported to understand if there are regression we might have missed.

Sam on behalf of the React Team.

**Reply by psionic12** [reply]
Hi,
May I ask what is "neutral"? Does this mean that the problem (new arch may not as fast as the old arch) is known and acceptable, as long as the gap is not noticeable?

**Reply by Almouro** [reply]
Hi @sammy-SC, thanks a lot for your answer! 🙏 

Your last paragraph is very nice to hear! 
I'm guessing that by neutral you mean:
- either the performance change was negligible overall 
- or it was slightly better for some components and slightly worse for some others?

I agree that the memory increase seems ok (although Android TV devs having to run in very low RAM conditions might disagree), but it's good to know it's being worked on. 🥳 (also concurrent mode is no small feature of course!)

To answer your questions:
1. I've only run it on 2 separate devices (my own Samsung J3 2017 and AWS Samsung A10s). Both are low-end, `armeabi-v7a` abi
2. No real reason for Device Farm except laziness 😅, easier to run measures in the cloud than on my own devices
3. Never tried iOS mostly because I'm generally less concerned with iOS performance since the devices usually have much better CPU

Thanks for taking the time to try and reproduce this out, I really appreciate it! 

Interesting to see that you're getting different results, would you mind sharing your json of measures for both scenarios?


I haven't dived deep into profiling yet, but it's worth noting that for texts, the major difference seemed to come from the `ReactTextViewManager.updateState` method being called while `updateState` was never called for the views

**Reply by sammy-SC** [reply]
@psionic12 

> Hi, May I ask what is "neutral"? Does this mean that the problem (new arch may not as fast as the old arch) is known and acceptable, as long as the gap is not noticeable?

By neutral I mean the difference between old and new architecture is not statistically significant. It is not whether the difference is perceivable by the end user but if the difference is measurable. It does not mean that the new architecture is not as fast as the old architecture.

**Reply by sammy-SC** [reply]
@Almouro 

Here are the JSONs used to generate the screenshot above: https://gist.github.com/sammy-SC/8c2e589894c12048ff45a35b77ba5aac


> I'm guessing that by neutral you mean:
>  - either the performance change was negligible overall
>  - or it was slightly better for some components and slightly worse for some others?

I ran your benchmarks several times but I was unable to replicate your results. For example "showing 100 tweets" benchmark would not consistently show a win in one architecture or the other. That's what I meant to say when I said "neutral" in context of running your benchmarks.
At Meta, we were not measuring how long it takes to draw individual components. We don't know if performance regression in one place was offset by perf win elsewhere. The focus was on entire screens and how long it takes to paint them. Performance change overall was negligible (often referred to as not statistically significant).

**Reply by Almouro** [reply]
Hi @sammy-SC! Thanks for your answer, and for providing the measures! 🙏

Apologies for the long message 😅
(btw I'm also available on the RN contributors discord  to discuss further if you prefer it!)

- Quick question first, what were the devices for which the Facebook app performance impact was tested? (or was it data from production users?)

## Checking your results

Worth noting here that the Pixel 3 and Pixel 5 are quite high-end, so if there is in fact a performance impact, it might be less noticeable.
Even so, if we reduce the measures you provided to a duration of 3s since both versions finish in ~3s on your device, [this is the result](https://rn-new-arch-perf.netlify.app/pixel3/100tweets/report). 
They're quite similar, but I'm still seeing the increased TTI and a 27% overhead in UI thread CPU consumption.

![image](https://user-images.githubusercontent.com/4534323/198647575-a0963761-021f-48b2-9d0b-e1ccc7d85e26.png)

## New measures with high end device

My hypothesis is that with more tweets you'd see the impact on the high-end phone
If you want to try it out, I've added 
- APKs with 300 tweets ([here](https://github.com/Almouro/fabric-test/blob/main/apks/fabric-300-tweets.apk) and [here](https://github.com/Almouro/fabric-test/blob/main/apks/nofabric-300-tweets.apk))
- APKs with 10k texts ([here](https://github.com/Almouro/fabric-test/blob/main/apks/fabric-10k-texts.apk) and [here](https://github.com/Almouro/fabric-test/blob/main/apks/nofabric-10k-texts.apk))

I've tried those out on my Samsung S10 (even higher end): 
- [300 tweets](https://rn-new-arch-perf.netlify.app/s10/300tweets/report) 
  - impact seems low, still TTI is up ~10%, UI thread up 12%
- [10k texts](https://rn-new-arch-perf.netlify.app/s10/10k-texts/report)
  - again we see like below the big impact for text rendering, mainly coming from heavier thread usage
![image](https://user-images.githubusercontent.com/4534323/198655546-2267fb4a-3f21-465c-a950-576bc8522a34.png)

## Choice of test devices

Using Geekbench CPU benchmarks for example, here is a comparison of CPU performance for all the devices used in this conversation so far:
(I'm also adding my own S10 on which I ran some new measures)

Device | [Pixel 5](https://browser.geekbench.com/android_devices/google-pixel-5) | [Pixel 3](https://browser.geekbench.com/android_devices/google-pixel-3) | [J3](https://browser.geekbench.com/android_devices/samsung-galaxy-j3-2017) | [A10s](https://browser.geekbench.com/android_devices/samsung-galaxy-a10s) | [S10](https://browser.geekbench.com/android_devices/samsung-galaxy-s10-qualcomm-snapdragon-855-1-8-ghz) |
| - | --- | ---| ---| -| -|
| Single-core score |570 | 448 | 104 | 135 | 681 |
| Multi-core score |1533 | 1629 | 354 | 480 | 2452 |

The reason for using such low end devices for testing is that we found recently that a lot of our users had quite low end devices, but I imagine the FB app would be quite similar.

Based on that [small research](https://blog.bam.tech/developer-news/what-is-the-performance-of-your-users-devices), **roughly 25% of our Android users would basically have a worse phone than the A10s I'm testing on**.
This is not even taking into account our Android TV projects which have to face even crappier devices 😅

## About Flatlist

I might have missed something here, but Flatlist seems to be consistently performing worse on low end and high end Android phones. Even on my S10, scroll locks which seems to correlate with heavy `fabric_bg` usage [report](https://rn-new-arch-perf.netlify.app/s10/flatlist/report) and [code + APKs](https://github.com/Almouro/fabric-test#scenario-5-pokedex-flatlist)

It's true that on the old arch, Flatlist is not the best component anymore, and I think internally FB uses something else?

But the thing is the new Flashlist is struggling with performance with the new architecture, seemingly for similar reasons https://github.com/Shopify/flash-list/pull/550#issuecomment-1292068129

On the issue, @j-piasecki shares a system trace on which we see `updateState` for texts being perhaps the bottleneck
![image](https://user-images.githubusercontent.com/4534323/198659229-8e2196d5-2848-4402-8d65-d4129bcc2c04.png)

## My current conclusions and concerns

- Of course, no apps should be rendering thousands of components. I also understand performance gains should not necessarily be expected especially for such simple scenarios (I'm not even taking advantage of concurrent rendering anywhere)

- But in most scenario closer to real life (for instance this one, https://github.com/Almouro/fabric-test#scenario-6-navigation), the new arch seems to be consistently performing a bit worse. Like 10% increased CPU usage and render time, which is not negligible.
On high end devices, the impact seems less clear indeed.
Mainly the impact seems to be on the UI thread, which is concerning to me, since this might mean low end devices struggle more to achieve 60FPS

- Might be wishful thinking, but maybe there's just something to be improved for Text or layouting?
When I see the fantastic improvement we have for simple views I do wish we'd have something similar for texts 🤞

**Reply by sammy-SC** [reply]
I'll answer in several parts to break up the discussion into pieces. I'm happy to talk on Discord as well but I feel like it is more accessible/searchable on Github.

> Quick question first, what were the devices for which the Facebook app performance impact was tested? (or was it data from production users?)

It was tested with data from real world users. Surfaces in Facebook app are instrumented so we can measure how long it takes from user initiating navigation until a marker element is rendered on screen. Collected data was grouped by device class: low end, mid end and high end to make sure all types of devices have neutral perf.

**Reply by sammy-SC** [reply]
Pixel 3 was the oldest found I could get my hands on the day I was re-running your benchmarks. Low end Android devices are important. As I mentioned in my previous comment, we were breaking down perf metrics by device class to make ensure perf neutrality for all device classes. The new architecture should not perform worse on low end devices than the old architecture. 

Where in the results do you see TTI? is it "Processes with high CPU usage detected"? 
Why do you think higher CPU UI thread utilisation is bad? The metric is about how much CPU is waiting for cache or spending time context switching. it is worth keeping track of it but I don't understand how it makes the experience worse for the end user.

**Reply by sammy-SC** [reply]
> Of course, no apps should be rendering thousands of components. I also understand performance gains should not necessarily be expected especially for such simple scenarios (I'm not even taking advantage of concurrent rendering anywhere)

Even though no app will render thousands of components, I think this is a good exercise. When I was sharing details about how we did rollout, the point was that we were looking at real world usage of the framework. Examples likes this can point to perf bugs, which is something I hope we can discover and fix.

> Might be wishful thinking, but maybe there's just something to be improved for Text or layouting?

This certainly can be the case. The new architecture uses lower level Yoga APIs. This means more responsibility is on side of RN than before. Previously, the Yoga tree was constructed through high level interface and everything was taken care of. There is a trade off between how much RN has control over things and simplicity of Yoga integration.
Also the way text is measured has changed. The critical path is in C++ and it could be that we do unnecessary copy of heavy objects or text measurements are not cached correctly.
I'll run the new benchmarks you prepared (thank you!) and see if I can get a consistent repro. When I have a repro I can dig in and see look at what's going on.

**Reply by Almouro** [reply]
Hi @sammy-SC, thanks for your answers! I really appreciate you going out of your way to reply 🙏

## TTI

in [this report](https://rn-new-arch-perf.netlify.app/pixel3/100tweets/report), I was referring to the TTI from the average run time, since the test just waits for element to appear.   
But **the measure isn't accurate, so it might not reflect the reality** ->  it's just measuring the Appium test execution and I've just realized that Appium waits for UI thread to be done to search for elements, so the measure might just be wrongly higher just because of the higher UI thread usage

But to get my TTI results, I was just manually timing the appearance of elements by comparing videos (like this video for 100 tweets on [low end devices](https://twitter.com/almouro/status/1583070070275985409). At least on low end devices, TTI was consistently ~20% worse.

## UI Thread

- Why do you think higher CPU UI thread utilisation is bad?

Since the UI thread is still ultimately responsible for orchestrating the frame rendering at 60FPS (or more), my understanding is that more pressure on the UI thread might make it harder to achieve perfect frame rate and induce higher risk of lag. 
Especially when rendering lots of elements (like in list scrolling), or for users with low end phones with bad CPU cores that might get saturated easily.

To be honest, it's also the correlation I could find with the higher TTI, but that was mainly guessing. 
When I'm taking a callstack trace with Android studio for the example with thousands of texts, the higher TTI seems to correlate with a few more seconds spent on the UI thread for `ReactTextViewManager.updateState`
@psionic12 seems to have found very interesting insights here (https://github.com/reactwg/react-native-new-architecture/discussions/88#discussioncomment-4044743): mainly that this might be linked to too much JNI calls to read from native maps

**Comment by sammy-SC** [comment]
Hello @Almouro,

we spent some time analysing performance and summarised our findings in https://github.com/reactwg/react-native-new-architecture/discussions/123.

If you have some time, could you please try to run your benchmarks on the [0.72.0-RC.1](https://github.com/facebook/react-native/releases/tag/v0.72.0-rc.1) and see what the performance is like? We've made some improvements.

**Comment by nateshmbhat** [comment]
any article discussing the RN 0.76 performance ?

**Comment by sarthak-d11** [comment]
https://reactnativebenchmark.dev/
This website benchmarks all react native versions new arch and old arch as well

**Reply by kelset** [reply]
that website and repo is kind of interesting! is there a blogpost or something detailing how it works, to get to know better?
